### ISSUE TITLE
Accessing database from a background Isolate rather than main Isolate

### ISSUE THREAD

**Issue by amugofjava** [tekartik/sqflite#1118]
Hi,

Firstly, thank you for your great packages. I've been using your Sembast for years and it has been rock solid and it's so easy to use. I am now looking to Sqlflite, as I thought you could use it to access a Sqlite database in the background, but I am unsure.

On the usage recommendation page, it says that access should only occur in the main Isolate; however, on the changelog, v2.2.0+3 says initial support of cross Isolate safe.

Please could you clarify if accessing a database from both the main Isolate, and a background Isolate (e.g. for writing records during a scheduled task) is possible and/or recommended. 

Apologies if I this is mentioned in a change or readme and I've missed it.

**Comment by alextekartik** [comment]
Indeed many people have asked about access from another isolate and unfortunately I don't have a good answer for that and the documentation is unclear about that, sorry.
It **should** work although I have not tested it much and I have avoided it (for example by saving the temporary data read
in a file that I handle later in the main isolate). One common usage
on Android and iOS is to have access from a push notification isolate or scheduled task.
* Good if you can avoid it and only limit to this case (background notification / scheduled task)
* In both isolates
  * Use `singleInstance: false` in `openDatabase`
  * Don't close the database
* Since hot reload might fail when using `singleInstance: false` (if done while a transaction is in progress), you 
  can try `singleInstance: true` in the main isolate only.

**Comment by ced1check** [comment]
I'm using this great package, and it works really great.

There is one area, caused by flutter architecture, where it is problematic, storing lots of data.

So I tried to move the loading function (from web server to DB) in an isolate, and faced many issues.

First I was not using sqflite_common_ffi at all, but kept getting this error in the isolate:

`When using sqflite_common_ffi You must call databaseFactory = databaseFactoryFfi;
`

Any way to avoid using ffi in the isolate? Cause I don't like how it works, insert don't return results and it seems to cause a lot of issues.


Now, I'm facing weird problems when attempting the first data write:

It reports this:

```
SqfliteFfiException(sqlite_error: 1032, , SqliteException(1032): while executing statement, attempt to write a readonly database, attempt to write a readonly database (code 1032)
...
{details: {database: {path: xxx.db, id: 2, readOnly: false, singleInstance: false}}}
```

Tried with singleInstance true or false to no avail.

However data is stored, so If i restart the whole process it ends-up loading more data, not in a reliable way though.

And with more errors like those:

```
Corruption reported by sqlite on database

SqfliteDatabaseException DatabaseException(attempt to re-open an already-closed object: SQLiteDatabase: xxx.db)
```

All this to say, I don't think the library is ready for use in a separate isolate.

**Comment by ced1check** [comment]
> * In both isolates
>   
>   * Use `singleInstance: false` in `openDatabase`
>   * Don't close the database
> * Since hot reload might fail when using `singleInstance: false` (if done while a transaction is in progress), you
>   can try `singleInstance: true` in the main isolate only.

I'm currently trying to use the plugin in an isolate to load data from web server and store it in DB.

@alextekartik Is there a way to initialize sqflite without ffi in the isolate like it is in the main isolate?

As you suggest I don't close the DB anywhere, except manually to delete it from app's login page.

Still I end-up with weird errors mentioned above, could it be because I use ffi in isolate and not in main?

**Comment by ced1check** [comment]
Yet again I wrote too fast. Using isolates brings a lot of issues on Android.

Using sqflite in isolate is bound to break/corrupt DB and even lock it for several minutes!?

Using sqlite3 in isolate works without concurrency issues but is not reliable !? Amongst 50+ tables, 2 tables do not receive their data for some obscure reasons.

I'm still fiddling with settings (keeping DB open, opening single instance, checking inserted data, etc...) but it's time consuming.


Interestingly (or frustratingly) there is no UI freezing on iOS and no need for isolates, big DB inserts takes milliseconds instead of several seconds. I don't understand why there is such a big difference: inserting 4000 items on Android takes 2 seconds on Pixel 6a, 200ms on iPhone 12. Why???

I'm feeling lonely with this issue, so if someone has any hints or ideas that could help, I'll take it.

**Comment by alextekartik** [comment]
Sorry for the trouble you're hitting. As I said above, that is why I avoid using it from other isolate.

>  I don't understand why there is such a big difference: inserting 4000 items on Android takes 2 seconds on Pixel 6a, 200ms on iPhone 12. Why???

Performance-wise, this might be due to WAL not enabled by default on Android when using sqflite. You can try the following trick: https://github.com/tekartik/sqflite/blob/master/sqflite/doc/dev_tips.md#enable-wal-on-android

sqflite_common_ffi (which uses sqlite3 in a separate isolate) should give better performance than sqflite and you should be able to turn WAL mode on in a classic SQL way (`PRAGMA journal_mode=WAL;`) during onConfigure.

For insert I'm assuming you're using transaction or batch. 

You could also consider trying https://pub.dev/packages/sqflite_common_ffi_async which use sqlite3_async with WAL enabled by default. Changing between sqflite, sqflite_common_ffi and sqflite_common_ffi_async is just a matter of changing the database factory, other code remains the same.

**Comment by ced1check** [comment]
Thanks for your reply, much appreciated.

I tried WAL and FFI before, and only noticed slight performance improvements.

Worth noting about FFI, it doesn't report row IDs.


Got an interesting insight here: https://github.com/simolus3/sqlite3.dart/issues/296
You will see more precise timing differences between Android and iOS there by the way, when using plain INSERT, no parameters.

Turns out it' way better to use parameterized execute(stmt, values) or prepared statement with parameters, no plain SQL, huge difference on Android at least.

Moving sqflite code to do the same greatly reduced timing to under 200-300ms. which is way more reasonable.

Yet to compare that new code with iOS to see if statement compilation is still slower on Android than iOS as suggest by simolus3 on in sqlite3 thread.



Long story short: use prepared statement and/or execute(parameterized_sql, List<Object>) as much as possible.

I feel stupid not having followed that route initially as I had that list<object> ready in the first place! Probably too lazy calculating the initial statement with the right amount of question mark (?,?,?...) LOL.


I'm still trying to use isolates because transforming the received json data takes time but it seems another complicated area, where sometimes things get out of hand. I might open a new "issue" to report on that.

**Comment by alextekartik** [comment]
Interesting indeed. And sqflite does not offer prepared statement so you might have to stick with sqlite3.

**Comment by ced1check** [comment]
In sqflite, using execute(INSERT xxxx (fields) values (?,?,?...) seems to do better than using a plain SQL with all values already filled in! Timing has drop from 1+ second to 300ms, which is not that bad.

That said, I've successfully mixed sqflite and sqlite3 (no separate isolate) by keeping DB open on both and it appears reliable. 

Using a separate isolate bring very odd issues, but I might try again later.

**Comment by ced1check** [comment]
Turns out that mixing sqflite and sqlite3 on iOS is not working reliably whereas Android is ok with that.

One should be careful during DB upgrade though, if upgrade is performed on sqflite, sqlite3 access will not wait for the ugprade to complete.


Overall sqflite shows an overhead of about 200ms storing about 4000 rows, possibly due to the fact that it doesn't have prepared statements.


And one should use parameterized insert SQL instead of plain insert SQL regardless of plugin used:

This is extremely slow on Android:
`INSERT table (field1...fieldX) values (value1...valueX)`

This is ok on both Android/iOS:
`INSERT table (field1...fieldX) values (?...?)`

**Comment by ced1check** [comment]
FWIW I tried to switch to sqlite3 only, but loading data went from 2 seconds to 18 seconds on Android, 300ms to 2 seconds on iOS :(

I suppose transforming sqlite3 ResultSet  to List<Map<String, Object>> is better to be done in native (like sqflite is doing I suppose) than flutter.

That's the code I used, not sure if there's better way to do this?

```
      var stmt = db!.prepare(sql);
      ResultSet rs = stmt.select(whereArgs ?? []);
      stmt.dispose();

      List<Map<String, Object?>> list = [];

      for(List<Object?> row in rs.rows) {
        int idx = 0;
        Map<String, Object?> object = {};
        for(Object? obj in row) {
          object[ rs.columnNames[idx++] ] = obj;
        }
        list.add(object);
      }
```
